### DISCUSSION TITLE
Add a way to define compound indexes/indexes with partialFilterExpression/etc

### DISCUSSION THREAD

**Discussion by AlexeyBond** [payloadcms/payload#1121] [Feature Requests & Ideas]
We're currently considering migration of a relatively huge (few millions of documents) database from Strapi to Payload.
The way Payload works with MongoDB seems to be more efficient (especially in case of relations and nested documents).
However there are some issues that may become critical.

One of issues that don't seem to be solvable without modifying Payload's source code is - there is no way to define a index with a `partialFilterExpression`, on which we rely in some cases.
It would also be nice to be able to define compound indexes in code.

I suggest to extend collection/field configs to support like one in the following example:
```TypeScript
const MyCollection: CollectionConfig = {
  fields: [
    ...,
    {
      name: 'foo',
      type: 'text',
      ...,
      // along with a boolean values accept objects containing index settings
      // looks like this even may work now, except there is no support in typings
      index:  {
        unique: true,
        partialFilterExpression: {
          foo: /.+/
        },
      },
    },
  ],
  ...,
  // optional array of settings for additional indexes, including compound
  indexes: [
    {
      fields: { foo: 1, bar: -1 },
      options: { unique: true },
    },
  ],
};
```

**Comment by DanRibbens** [comment]
This is a great feature request! I have thought about this before actually and was waiting for somebody to ask.

I can definitly see us doing that and I like your suggestion on how to extend the config. If you are interested in contributing we would **gladly** accept a PR. We do all our indexes in `src/mongoose/buildSchema.ts`.

We can give more direction too, let us know. 
Thanks!

**Reply by AlexeyBond** [reply]
Hi. Thanks for the response!

I think I will be able to implement this feature if migration of our project to Payload will be approved by our managers.

**Comment by thanaParis** [comment]
Was there ever any progress on this? @DanRibbens ? Facing a similar issue of moving from one CMS to another and finding a lack of compound index/uniqueness support in Payload CMS.

**Comment by nDaniel86** [comment]
I am also waiting for the release of this feature in the CMS

**Comment by delkant** [comment]
+1 any plans on implementing this?

**Comment by SVronskiy** [comment]
+1

**Comment by SVronskiy** [comment]
Compound indexes already realized in hook afterSchemaInit. You can read about it here:
https://payloadcms.com/docs/database/postgres#afterschemainit
https://payloadcms.com/docs/database/sqlite#afterschemainit

**Comment by christophemenager** [comment]
https://payloadcms.com/docs/database/indexes#compound-indexes

**Comment by dolanoadriano** [comment]
I think that `partialFilterExpression` should also be implemented.

@r1tsuu
